### ISSUE TITLE
Switch from github.com/pkg/browser to github.com/cli/browser

### ISSUE THREAD

**Issue by mikelolasagasti** [opentofu/opentofu#603]
### OpenTofu Version

```shell
main
```

### Use Cases

https://github.com/pkg/browser/ has seen little activity since 2021 and GitHub forked it in https://github.com/cli/browser/ and is used by their CLI. The fork has better support for WSL and Windows environments.

### Attempted Solutions

-

### Proposal

Switch from github.com/pkg/browser to github.com/cli/browser

**Comment by cube2222** [comment]
Thanks for creating the issue @mikelolasagasti!

This makes sense to me, though I'm a bit worried that the fork (cli/browser) has issues disabled on github, so it's a bit hard to gauge if there are bugs or not.

Anyway, I think we'll have to wait with this change until after the stable release.

**Comment by serdardalgic** [comment]
Just out of curiosity, I was checking these repositories and bumped into[ this comment](https://github.com/pkg/browser/issues/54#issuecomment-1617677851):

> Imho, even cli/browser isn't actively developed. Because they [refuse to upgrade dependencies](https://github.com/cli/browser/pull/6#issuecomment-1428189439). Also [here](https://github.com/cli/browser/pull/5#discussion_r1104439895)

Still, I think switching from `pkg/browser` to `cli/browser` is a better option. But it may not be the final step in fixing the problem. So, +1 to @cube2222 's concerns.

Digging deeper, the OpenURL is only used during login [here](https://github.com/opentofu/opentofu/blob/main/internal/command/login.go#L482) and one above method is interactiveGetTokenByCode  which is called [here](https://github.com/opentofu/opentofu/blob/main/internal/command/login.go#L188).

**Comment by davidjeddy** [comment]
Not sure I have enough context / experience to add anything meaningful to this conversation; but I am a gluten for punishment so here goes:

`...I'm a bit worried that the fork (cli/browser) has issues disabled on github...` This greatly concerns me. Not being willing to accept issues the community/users have found means the package owner has little to no concern for managing the projects future.

`...they [refuse to upgrade dependencies](https://github.com/cli/browser/pull/6#issuecomment-1428189439)...` My gut says avoid. Being unwilling to stay current because `nothing is broken` makes me not trust the publisher willingness to address future issues.

In my opinion stability and maintainability > performance.

**Comment by Yantrio** [comment]
It's worth noting that they have since updated dependencies:
https://github.com/cli/browser/commit/708f5e216d0e45e1c7591d4c1a226ad4e9ab08aa

I think we should switch to their latest version. Even if they have issues disabled.

**Comment by mikelolasagasti** [comment]
Updated PR to latest 1.3.0 version.

**Comment by williammartin** [comment]
Hey folks, I'm a GitHub CLI maintainer and I just happened to stumble across this thread while doing some `cli/browser` housekeeping. Although I wasn't part of the team initially, and no one who made the decision is still around, I can probably shed some light on a few things. Sorry to @davidjeddy for quoting your specific sections, my intention is not to "call you out", your comment just touched on a number of topics that seem useful to address.

> Not being willing to accept issues the community/users have found means the package owner has little to no concern for managing the projects future.

We are concerned with the future of `cli/browser` to the extent that it fulfils the needs of `cli/cli`. In terms of issues being turned off, I don't think the team ever really wanted to take maintenance over a browser package beyond issues that were present in the CLI itself; we're a small team (sits at around 3 engineers, though currently 2). There's obviously a trade off in this decision and I'd be open to discussing this, maybe even opening issues, but by itself that wouldn't fundamentally change our view on what we'd want to accept, or our capacity to do so. What it **would do** is allow an opportunity for people to at least raise issues that we aren't aware of, and maybe we can strike a balance with a clear policy about what we'd accept.

> Being unwilling to stay current because nothing is broken makes me not trust the publisher willingness to address future issues.

I don't think that's really the position Mislav was taking in the linked PRs. In the case of https://github.com/cli/browser/pull/5#discussion_r1104439895, this was a backwards incompatible change for anyone not using `1.16+` due to the use of `os.CreateTemp` and in the case of https://github.com/cli/browser/pull/6 the same was true for `1.17+` due to `unsafe.Slice`. So with regards "stability and maintainability > performance." (though I'm not actually sure what performance you mean here), Mislav was valuing stability for anyone using the module; it's not quite so clear cut as the inequality makes it out to be!

In any case, as there was a vulnerability in `golang.org/x/sys` ([I think this one, though I may not recall correctly](https://github.com/advisories/GHSA-p782-xgp4-8hr8)), it seems like Sam decided to bite the bullet and move everything forward. I'm not sure going all the way to `1.21` was strictly necessary but here we are and we're no longer an exact drop in replacement for `pkg/browser` so that's out of the window anyway. I don't think that's the worst thing to happen, and maybe it opens up some other opportunities.

---

TL;DR:

I'm open to having a discussion about the maintenance of `cli/browser` if someone wants to have it. I have a loosely held opinion that we wouldn't change our position much around maintenance but at the very least then expectations would be set correctly, and that seems valuable for y'all who are currently consuming the module and anyone else making a decision in the future.

Cheers all!

**Comment by davidjeddy** [comment]
@williammartin Thank you for the clearification and context. The additional understand goes a long ways towards finding solutions that work for all project maintaners.

**Comment by apparentlymart** [comment]
Hi again, all!

The OpenTofu maintainers discussed this some more last week and we noted that OpenTofu's needs for launching web browsers are very modest -- it is currently purely for `tofu login` to launch an interactive OAuth consent flow -- and so we're unlikely to need anything more than GitHub CLI needs.

Furthermore, `tofu login` interacts with `github.com/cli/browser` through a lightweight internal abstraction anyway and so we can could swap it out for any similar library later if we need to.

Therefore we think it's pretty unlikely that we'd actually need any new functionality nor specific fixes from the upstream library that would cause additional maintenance burden for the team at GitHub. If we do end up needing more than what GitHub CLI needs then we'd be happy to discuss co-maintaining if the team is open to it, or alternatively switch to our own copy of the same code if what we need is far beyond the scope that GitHub CLI needs. But again, we think that's highly unlikely.

With all of that said then: we intend to keep using `github.com/cli/browser` for now, and consider its functionality already sufficient for what OpenTofu currently needs. We'll revisit this if the situation changes, but we don't have any current plans that would require more than what this library already provides.
